# Supervised Learning using Linear Regression

## Objective
- Determine a process of building linear regression model in big data

## Hypothesis
- Building a linear regression model in big data requires formal definition of a target

## Procedure

In [ ]:
# Run this cell
!pip install pyspark
!pip install findspark

In [ ]:
# Run this cell
!pip install awscli

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 4.5/4.5 MB 44.7 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 12.6/12.6 MB 90.1 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 548.2/548.2 kB 33.1 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 82.6/82.6 kB 6.4 MB/s eta 0:00:00
  Attempting uninstall: rsa
    Found existing installation: rsa 4.9
    Uninstalling rsa-4.9:
      Successfully uninstalled rsa-4.9
  Attempting uninstall: docutils
    Found existing installation: docutils 0.18.1
    Uninstalling docutils-0.18.1:
      Successfully uninstalled docutils-0.18.1


In [ ]:
# Run this cell
import matplotlib.pyplot as plt
import numpy as np

In [ ]:
# Run this cell
import findspark
findspark.init()
from pyspark.sql import SparkSession

spark = SparkSession.builder \
        .master('local[*]') \
        .appName('Basics') \
        .getOrCreate()

print(spark.version)

3.5.3


In [ ]:
# Run this cell
!mkdir raw_2178
!ls

raw_2178  sample_data


In [ ]:
# Run this cell
!aws s3 cp --recursive --no-sign-request s3://openaq-data-archive/records/csv.gz/locationid=2178/ raw_2178

download: s3://openaq-data-archive/records/csv.gz/locationid=2178/year=2016/month=03/location-2178-20160313.csv.gz to raw_2178/year=2016/month=03/location-2178-20160313.csv.gz
download: s3://openaq-data-archive/records/csv.gz/locationid=2178/year=2016/month=03/location-2178-20160310.csv.gz to raw_2178/year=2016/month=03/location-2178-20160310.csv.gz
download: s3://openaq-data-archive/records/csv.gz/locationid=2178/year=2016/month=03/location-2178-20160317.csv.gz to raw_2178/year=2016/month=03/location-2178-20160317.csv.gz
download: s3://openaq-data-archive/records/csv.gz/locationid=2178/year=2016/month=03/location-2178-20160306.csv.gz to raw_2178/year=2016/month=03/location-2178-20160306.csv.gz
download: s3://openaq-data-archive/records/csv.gz/locationid=2178/year=2016/month=03/location-2178-20160312.csv.gz to raw_2178/year=2016/month=03/location-2178-20160312.csv.gz
download: s3://openaq-data-archive/records/csv.gz/locationid=2178/year=2016/month=03/location-2178-20160315.csv.gz to ra

In [ ]:
# Run this cell
df_2178 = spark.read.csv('/content/raw_2178/*/*/', inferSchema=True, header=True)
df_2178.show(5)

+-----------+----------+--------------+--------------------+-------+-----------+---------+-----+--------------------+
|location_id|sensors_id|      location|            datetime|    lat|        lon|parameter|units|               value|
+-----------+----------+--------------+--------------------+-------+-----------+---------+-----+--------------------+
|       2178|      3916|Del Norte-2178|2023-12-29T01:00:...|35.1353|-106.584702|      no2|  ppm|              0.0327|
|       2178|      3916|Del Norte-2178|2023-12-29T02:00:...|35.1353|-106.584702|      no2|  ppm|0.033100000000000004|
|       2178|      3916|Del Norte-2178|2023-12-29T03:00:...|35.1353|-106.584702|      no2|  ppm|              0.0329|
|       2178|      3916|Del Norte-2178|2023-12-29T04:00:...|35.1353|-106.584702|      no2|  ppm|              0.0298|
|       2178|      3916|Del Norte-2178|2023-12-29T05:00:...|35.1353|-106.584702|      no2|  ppm|              0.0167|
+-----------+----------+--------------+-----------------

In [ ]:
# Run this cell
df_2178.count()

326544

In [ ]:
# Run this cell
df_2178.printSchema()

root
 |-- location_id: string (nullable = true)
 |-- sensors_id: string (nullable = true)
 |-- location: string (nullable = true)
 |-- datetime: string (nullable = true)
 |-- lat: string (nullable = true)
 |-- lon: string (nullable = true)
 |-- parameter: string (nullable = true)
 |-- units: string (nullable = true)
 |-- value: string (nullable = true)



In [ ]:
# Run this cell
df_2178.createOrReplaceTempView('df_2178_view')

In [ ]:
# Run this cell
spark.sql('''
SELECT
  *
FROM
  df_2178_view
LIMIT
  5
''').show()

+-----------+----------+--------------+--------------------+-------+-----------+---------+-----+--------------------+
|location_id|sensors_id|      location|            datetime|    lat|        lon|parameter|units|               value|
+-----------+----------+--------------+--------------------+-------+-----------+---------+-----+--------------------+
|       2178|      3916|Del Norte-2178|2023-12-29T01:00:...|35.1353|-106.584702|      no2|  ppm|              0.0327|
|       2178|      3916|Del Norte-2178|2023-12-29T02:00:...|35.1353|-106.584702|      no2|  ppm|0.033100000000000004|
|       2178|      3916|Del Norte-2178|2023-12-29T03:00:...|35.1353|-106.584702|      no2|  ppm|              0.0329|
|       2178|      3916|Del Norte-2178|2023-12-29T04:00:...|35.1353|-106.584702|      no2|  ppm|              0.0298|
|       2178|      3916|Del Norte-2178|2023-12-29T05:00:...|35.1353|-106.584702|      no2|  ppm|              0.0167|
+-----------+----------+--------------+-----------------

In [ ]:
# Run this cell
spark.sql('''
WITH
distinct_parameter AS (
SELECT
  DISTINCT parameter
FROM
  df_2178_view
)
SELECT
  *
FROM
  distinct_parameter
''').show()

+---------+
|parameter|
+---------+
|      so2|
|       co|
|      nox|
|       o3|
|     pm10|
|      no2|
|       no|
|     pm25|
+---------+



In [ ]:
# Run this cell
spark.sql('''
WITH
year_month_day_split AS (
SELECT
  *
  ,YEAR(datetime) AS year
  ,LPAD(MONTH(datetime), 2, 0) AS month
  ,LPAD(DAY(datetime), 2, 0) AS day
  ,LPAD(WEEKOFYEAR(datetime), 2, 0) AS weeknum
FROM
  df_2178_view
)
,year_month_day_cte AS (
SELECT
  *
  ,CONCAT(year, '-', month) AS year_month
  ,CONCAT(year, '-', month, '-', day) AS year_month_day
  ,CONCAT(year, '-', weeknum) AS year_weeknum
FROM
  year_month_day_split
)
,monthly_average AS (
SELECT
  year_month
  ,AVG(CASE WHEN parameter IN ('so2') THEN value ELSE NULL END) AS so2_mo_avg
  ,AVG(CASE WHEN parameter IN ('co') THEN value ELSE NULL END) AS co_mo_avg
  ,AVG(CASE WHEN parameter IN ('nox') THEN value ELSE NULL END) AS nox_mo_avg
  ,AVG(CASE WHEN parameter IN ('o3') THEN value ELSE NULL END) AS o3_mo_avg
  ,AVG(CASE WHEN parameter IN ('pm10') THEN value ELSE NULL END) AS pm10_mo_avg
  ,AVG(CASE WHEN parameter IN ('no2') THEN value ELSE NULL END) AS no2_mo_avg
  ,AVG(CASE WHEN parameter IN ('no') THEN value ELSE NULL END) AS no_mo_avg
  ,AVG(CASE WHEN parameter IN ('pm25') THEN value ELSE NULL END) AS pm25_mo_avg
FROM
  year_month_day_cte
GROUP BY
  year_month
)
SELECT
  *
FROM
  monthly_average
LIMIT
  5
''').show()

+----------+--------------------+-------------------+--------------------+--------------------+------------------+--------------------+--------------------+------------------+
|year_month|          so2_mo_avg|          co_mo_avg|          nox_mo_avg|           o3_mo_avg|       pm10_mo_avg|          no2_mo_avg|           no_mo_avg|       pm25_mo_avg|
+----------+--------------------+-------------------+--------------------+--------------------+------------------+--------------------+--------------------+------------------+
|   2024-09|3.687411598302690...|0.21450924608819338|0.005224550898203595| 0.04195977011494253|22.256373937677054|0.004887520525451561|0.001014970059880...| 5.778470254957506|
|   2024-02|                 0.0| 0.2645264847512038|0.013666144200626952|0.030589062499999993|15.198198198198199|  0.0104687106918239|0.003308777429467...|3.6265765765765754|
|   2023-12|                 0.0|0.29436416184971115|0.024935483870967733|0.019736918604651158| 15.52991452991453|0.0157

In [ ]:
# Run this cell
df_mo = spark.sql('''
WITH
year_month_day_split AS (
SELECT
  *
  ,YEAR(datetime) AS year
  ,LPAD(MONTH(datetime), 2, 0) AS month
  ,LPAD(DAY(datetime), 2, 0) AS day
  ,LPAD(WEEKOFYEAR(datetime), 2, 0) AS weeknum
FROM
  df_2178_view
)
,year_month_day_cte AS (
SELECT
  *
  ,CONCAT(year, '-', month) AS year_month
  ,CONCAT(year, '-', month, '-', day) AS year_month_day
  ,CONCAT(year, '-', weeknum) AS year_weeknum
FROM
  year_month_day_split
)
,monthly_average AS (
SELECT
  year_month
  ,AVG(CASE WHEN parameter IN ('so2') THEN value ELSE NULL END) AS so2_mo_avg
  ,AVG(CASE WHEN parameter IN ('co') THEN value ELSE NULL END) AS co_mo_avg
  ,AVG(CASE WHEN parameter IN ('nox') THEN value ELSE NULL END) AS nox_mo_avg
  ,AVG(CASE WHEN parameter IN ('o3') THEN value ELSE NULL END) AS o3_mo_avg
  ,AVG(CASE WHEN parameter IN ('pm10') THEN value ELSE NULL END) AS pm10_mo_avg
  ,AVG(CASE WHEN parameter IN ('no2') THEN value ELSE NULL END) AS no2_mo_avg
  ,AVG(CASE WHEN parameter IN ('no') THEN value ELSE NULL END) AS no_mo_avg
  ,AVG(CASE WHEN parameter IN ('pm25') THEN value ELSE NULL END) AS pm25_mo_avg
FROM
  year_month_day_cte
GROUP BY
  year_month
)
SELECT
  *
FROM
  monthly_average
''')

In [ ]:
# Run this cell
df_mo.show()

+----------+--------------------+-------------------+--------------------+--------------------+------------------+--------------------+--------------------+------------------+
|year_month|          so2_mo_avg|          co_mo_avg|          nox_mo_avg|           o3_mo_avg|       pm10_mo_avg|          no2_mo_avg|           no_mo_avg|       pm25_mo_avg|
+----------+--------------------+-------------------+--------------------+--------------------+------------------+--------------------+--------------------+------------------+
|   2024-09|3.687411598302690...|0.21450924608819338|0.005224550898203595| 0.04195977011494253|22.256373937677054|0.004887520525451561|0.001014970059880...| 5.778470254957506|
|   2024-02|                 0.0| 0.2645264847512038|0.013666144200626952|0.030589062499999993|15.198198198198199|  0.0104687106918239|0.003308777429467...|3.6265765765765754|
|   2023-12|                 0.0|0.29436416184971115|0.024935483870967733|0.019736918604651158| 15.52991452991453|0.0157

In [ ]:
# Run this cell
df_mo_avg = spark.sql('''
WITH
year_month_day_split AS (
SELECT
  *
  ,YEAR(datetime) AS year
  ,LPAD(MONTH(datetime), 2, 0) AS month
  ,LPAD(DAY(datetime), 2, 0) AS day
  ,LPAD(WEEKOFYEAR(datetime), 2, 0) AS weeknum
FROM
  df_2178_view
)
,year_month_day_cte AS (
SELECT
  *
  ,CONCAT(year, '-', month) AS year_month
  ,CONCAT(year, '-', month, '-', day) AS year_month_day
  ,CONCAT(year, '-', weeknum) AS year_weeknum
FROM
  year_month_day_split
)
,monthly_average AS (
SELECT
  year_month
  ,AVG(CASE WHEN parameter IN ('so2') THEN value ELSE NULL END) AS so2_mo_avg
  ,AVG(CASE WHEN parameter IN ('co') THEN value ELSE NULL END) AS co_mo_avg
  ,AVG(CASE WHEN parameter IN ('nox') THEN value ELSE NULL END) AS nox_mo_avg
  ,AVG(CASE WHEN parameter IN ('o3') THEN value ELSE NULL END) AS o3_mo_avg
  ,AVG(CASE WHEN parameter IN ('pm10') THEN value ELSE NULL END) AS pm10_mo_avg
  ,AVG(CASE WHEN parameter IN ('no2') THEN value ELSE NULL END) AS no2_mo_avg
  ,AVG(CASE WHEN parameter IN ('no') THEN value ELSE NULL END) AS no_mo_avg
  ,AVG(CASE WHEN parameter IN ('pm25') THEN value ELSE NULL END) AS pm25_mo_avg
FROM
  year_month_day_cte
GROUP BY
  year_month
)
,monthly_average_with_year_month_day1 AS (
SELECT
   *
   ,CONCAT(year_month, '-01') AS year_month_day1
FROM
  monthly_average
)
,monthly_average_in_2023 AS (
SELECT
  *
FROM
  monthly_average_with_year_month_day1
WHERE
  year_month_day1 BETWEEN '2023-01-01' AND '2023-12-01'
)
SELECT
  *
FROM
  monthly_average_in_2023
''')

In [ ]:
# Run this cell
df_mo_avg.show()

+----------+----------+-------------------+--------------------+--------------------+------------------+--------------------+--------------------+-----------------+---------------+
|year_month|so2_mo_avg|          co_mo_avg|          nox_mo_avg|           o3_mo_avg|       pm10_mo_avg|          no2_mo_avg|           no_mo_avg|      pm25_mo_avg|year_month_day1|
+----------+----------+-------------------+--------------------+--------------------+------------------+--------------------+--------------------+-----------------+---------------+
|   2023-12|       0.0|0.29436416184971115|0.024935483870967733|0.019736918604651158| 15.52991452991453|0.015776064610866364|0.009239002932551315|6.336039886039887|     2023-12-01|
|   2023-11|       0.0|                0.3| 0.01706231003039513| 0.02377762039660056|16.946927374301676| 0.01259037267080745|0.004770516717325227|5.803072625698324|     2023-11-01|
|   2023-07|       0.0| 0.2137199434229136|0.004522151898734179| 0.04965346534653465|26.4991974

In [ ]:
# Run this cell
df_mo_avg.summary().show()

+-------+----------+----------+-------------------+--------------------+--------------------+------------------+--------------------+--------------------+------------------+---------------+
|summary|year_month|so2_mo_avg|          co_mo_avg|          nox_mo_avg|           o3_mo_avg|       pm10_mo_avg|          no2_mo_avg|           no_mo_avg|       pm25_mo_avg|year_month_day1|
+-------+----------+----------+-------------------+--------------------+--------------------+------------------+--------------------+--------------------+------------------+---------------+
|  count|        12|        12|                 12|                  10|                  12|                12|                  12|                  10|                10|             12|
|   mean|      NULL|       0.0| 0.2610643468837323|0.009011901212766692|0.037558749965941385| 19.22817221626432| 0.00801872612232349|0.002516225824638...| 5.160932250424959|           NULL|
| stddev|      NULL|       0.0|0.05750742893269811

In [ ]:
# Run this cell
imputer_input = [
    'so2_mo_avg'
    ,'co_mo_avg'
    ,'nox_mo_avg'
    ,'o3_mo_avg'
    ,'pm10_mo_avg'
    ,'no2_mo_avg'
    ,'no_mo_avg'
    ,'pm25_mo_avg'
  ]
imputer_output = [
     'so2_mo_avg_impute'
    ,'co_mo_avg_impute'
    ,'nox_mo_avg_impute'
    ,'o3_mo_avg_impute'
    ,'pm10_mo_avg_impute'
    ,'no2_mo_avg_impute'
    ,'no_mo_avg_impute'
    ,'pm25_mo_avg_impute'
  ]

assembler_input = imputer_output
assembler_output = 'assembler_output'

scaler_input = assembler_output
scaler_output = 'scaler_output'

In [ ]:
# Run this cell
from pyspark.ml.feature import VectorAssembler, Imputer, StandardScaler
from pyspark.ml.pipeline import Pipeline

imputer = Imputer(inputCols=imputer_input, outputCols=imputer_output)
assembler = VectorAssembler().setInputCols(assembler_input).setOutputCol(assembler_output)
scaler = StandardScaler(inputCol=scaler_input, outputCol=scaler_output)

pipeline = Pipeline(stages=[imputer, assembler, scaler])

In [ ]:
# Run this cell
pipeline_model = pipeline.fit(df_mo_avg)

In [ ]:
# Run this cell
df_pipeline = pipeline_model.transform(df_mo_avg)
df_pipeline.show(5)

+----------+----------+-------------------+--------------------+--------------------+------------------+--------------------+--------------------+-----------------+---------------+-----------------+-------------------+--------------------+--------------------+------------------+--------------------+--------------------+------------------+--------------------+--------------------+
|year_month|so2_mo_avg|          co_mo_avg|          nox_mo_avg|           o3_mo_avg|       pm10_mo_avg|          no2_mo_avg|           no_mo_avg|      pm25_mo_avg|year_month_day1|so2_mo_avg_impute|   co_mo_avg_impute|   nox_mo_avg_impute|    o3_mo_avg_impute|pm10_mo_avg_impute|   no2_mo_avg_impute|    no_mo_avg_impute|pm25_mo_avg_impute|    assembler_output|       scaler_output|
+----------+----------+-------------------+--------------------+--------------------+------------------+--------------------+--------------------+-----------------+---------------+-----------------+-------------------+----------------

In [ ]:
# Run this cell
from pyspark.ml.functions import vector_to_array

df_array = df_pipeline.withColumn('array_scaler_output', vector_to_array("scaler_output"))
df_array.show(5)

+----------+----------+-------------------+--------------------+--------------------+------------------+--------------------+--------------------+-----------------+---------------+-----------------+-------------------+--------------------+--------------------+------------------+--------------------+--------------------+------------------+--------------------+--------------------+--------------------+
|year_month|so2_mo_avg|          co_mo_avg|          nox_mo_avg|           o3_mo_avg|       pm10_mo_avg|          no2_mo_avg|           no_mo_avg|      pm25_mo_avg|year_month_day1|so2_mo_avg_impute|   co_mo_avg_impute|   nox_mo_avg_impute|    o3_mo_avg_impute|pm10_mo_avg_impute|   no2_mo_avg_impute|    no_mo_avg_impute|pm25_mo_avg_impute|    assembler_output|       scaler_output| array_scaler_output|
+----------+----------+-------------------+--------------------+--------------------+------------------+--------------------+--------------------+-----------------+---------------+------------

In [ ]:
# Run this cell
df_scale = df_array.withColumn('so2_mo_avg_scale', df_array['array_scaler_output'].getItem(0))\
.withColumn('co_mo_avg_scale', df_array['array_scaler_output'].getItem(1))\
.withColumn('nox_mo_avg_scale', df_array['array_scaler_output'].getItem(2))\
.withColumn('o3_mo_avg_scale', df_array['array_scaler_output'].getItem(3))\
.withColumn('pm10_mo_avg_scale', df_array['array_scaler_output'].getItem(4))\
.withColumn('no2_mo_avg_scale', df_array['array_scaler_output'].getItem(5))\
.withColumn('no_mo_avg_scale', df_array['array_scaler_output'].getItem(6))\
.withColumn('pm25_mo_avg_scale', df_array['array_scaler_output'].getItem(7))

df_scale.show(5)

+----------+----------+-------------------+--------------------+--------------------+------------------+--------------------+--------------------+-----------------+---------------+-----------------+-------------------+--------------------+--------------------+------------------+--------------------+--------------------+------------------+--------------------+--------------------+--------------------+----------------+------------------+------------------+------------------+------------------+------------------+------------------+-----------------+
|year_month|so2_mo_avg|          co_mo_avg|          nox_mo_avg|           o3_mo_avg|       pm10_mo_avg|          no2_mo_avg|           no_mo_avg|      pm25_mo_avg|year_month_day1|so2_mo_avg_impute|   co_mo_avg_impute|   nox_mo_avg_impute|    o3_mo_avg_impute|pm10_mo_avg_impute|   no2_mo_avg_impute|    no_mo_avg_impute|pm25_mo_avg_impute|    assembler_output|       scaler_output| array_scaler_output|so2_mo_avg_scale|   co_mo_avg_scale|  nox_mo

In [ ]:
# Run this cell
from pyspark.ml.regression import LinearRegression

assembler = VectorAssembler(inputCols=['co_mo_avg_scale', 'nox_mo_avg_scale', 'o3_mo_avg_scale'], outputCol='features')
df_assembled = assembler.transform(df_scale)

estimator = LinearRegression(featuresCol='features', labelCol='pm10_mo_avg_impute')

model = estimator.fit(df_assembled)

df_prediction = model.transform(df_assembled)

In [ ]:
# Run this cell
df_prediction.show(5)

+----------+----------+-------------------+--------------------+--------------------+------------------+--------------------+--------------------+-----------------+---------------+-----------------+-------------------+--------------------+--------------------+------------------+--------------------+--------------------+------------------+--------------------+--------------------+--------------------+----------------+------------------+------------------+------------------+------------------+------------------+------------------+-----------------+--------------------+------------------+
|year_month|so2_mo_avg|          co_mo_avg|          nox_mo_avg|           o3_mo_avg|       pm10_mo_avg|          no2_mo_avg|           no_mo_avg|      pm25_mo_avg|year_month_day1|so2_mo_avg_impute|   co_mo_avg_impute|   nox_mo_avg_impute|    o3_mo_avg_impute|pm10_mo_avg_impute|   no2_mo_avg_impute|    no_mo_avg_impute|pm25_mo_avg_impute|    assembler_output|       scaler_output| array_scaler_output|so2_

In [ ]:
# Run this cell
from pyspark.ml.evaluation import RegressionEvaluator

evaluator = RegressionEvaluator(labelCol='pm10_mo_avg_impute', predictionCol='prediction', metricName='rmse')
rmse = evaluator.evaluate(df_prediction)

print('Root Mean Squared Error (RMSE): ', rmse)
print('Coefficients: ', model.coefficients)
print('Intercept: ', model.intercept)

Root Mean Squared Error (RMSE):  2.911344046880226
Coefficients:  [-0.177192925495847,1.680420269973981,4.16154948957664]
Intercept:  2.334707428584541


## Observation and Analysis

Based on my observations, there were important steps carried out in creating a linear regression model that predicts PM10 concentration from data available on air pollution, including data preparation, feature scaling, model training, and evaluation. First, missing values of air pollutants such as CO, NOx, O3, SO2, NO, PM25, and NO2 were addressed using an imputer and ensured clean data. The features are then aggregated using VectorAssembler, and StandardScaler is used afterwards to scale the feature values so that all of these features, or pollutants, are in the same range. I think that this is a good example of why preprocessing was important to this task because the magnitudes of the different pollutants varied highly and the model needs to be able to handle them properly.

First, the procedure have dealt with missing values for pollutants like CO, NOx, O3 and others, making sure that the data would be neat. Then it applied a technique, called VectorAssembler, to merge all the selected pollutants into one set of features that the model might use. It also applied StandardScaler to normalize the data; the procedure changed all the values to have the same scale so that the model treats every feature equally.

Then it fitted a linear model to predict the levels of PM10, using CO, NOx, and O3 as features. There were some interesting and not so intuitive relationships: an increase in CO lowered the PM10 level, whereas both NOx and O3 had a positive influence, that is to say, if these pollutants increased, then the PM10 would also increase, but for the highest influence on PM10 level, it was O3.

To see how good the model was at predictions, it computed the Root Mean Squared Error, which showed up as 2.911. RMSE is a measure that tells us how far off our predictions were from the true values; on average, the model's predictions were about 2.9 units away from the true PM10 values. It is not perfect, but it does show that the model can give an estimation of levels of PM10.

In general, it appears that O3 is the term that tends to dominate the prediction of the concentrations for PM10, and NOx contributes positively as well. This model would say that reducing the levels of PM10 might best be done by trying to reduce O3. A model like this performs fairly well, but it can be made much better by adding more features, discovering interactions between pollutants, or perhaps even a more complicated model may describe some of these more complicated relationships.

## Conclusion

Therefore, I conclude that this process manages to achieve the aim of developing a linear regression model in predicting PM10 from big data; in addition, it gives us excellent insights regarding the air pollutant relation. The steps taken include preparation of data, scaling of features, and modeling with a linear regression. With that process, we could predict the concentrations of PM10 using the levels of pollutants like CO, NOx, and O3.

The hypothesis is correct: the idea here is that developing a linear regression model for big data requires a formalized definition of a target in such a case, PM10 levels, and key features--air pollutants. So, the model established that setting PM10 as the target variable and CO, NOx, and O3 as features enables reasoning over PM10 concentrations. O3 had the largest coefficient and is the strongest predictor, so it leads to the conclusion that air pollutants indeed may be used as strong features for such models.

While the model performed moderately well with an RMSE of 2.911, there is room for improvement by incorporating additional features or using more advanced modeling techniques. Nonetheless, the procedure successfully built a basic predictive model that aligns with the initial hypothesis and objective.

# *Required Challenge*

Predict `o3_mo_avg_impute` using linear regression model. Use the following variables below as input to the model:
- `co_mo_avg_scale`
- `nox_mo_avg_scale`
- `pm10_mo_avg_scale`


In [ ]:
# Run this cell
from pyspark.ml.regression import LinearRegression

assembler = VectorAssembler(inputCols=['co_mo_avg_scale', 'nox_mo_avg_scale', 'pm10_mo_avg_scale'], outputCol='features')
df_assembled = assembler.transform(df_scale)

estimator = LinearRegression(featuresCol='features', labelCol='o3_mo_avg_impute')

model = estimator.fit(df_assembled)

df_prediction = model.transform(df_assembled)

In [ ]:
# Run this cell
df_prediction.show(5)

+----------+----------+-------------------+--------------------+--------------------+------------------+--------------------+--------------------+-----------------+---------------+-----------------+-------------------+--------------------+--------------------+------------------+--------------------+--------------------+------------------+--------------------+--------------------+--------------------+----------------+------------------+------------------+------------------+------------------+------------------+------------------+-----------------+--------------------+--------------------+
|year_month|so2_mo_avg|          co_mo_avg|          nox_mo_avg|           o3_mo_avg|       pm10_mo_avg|          no2_mo_avg|           no_mo_avg|      pm25_mo_avg|year_month_day1|so2_mo_avg_impute|   co_mo_avg_impute|   nox_mo_avg_impute|    o3_mo_avg_impute|pm10_mo_avg_impute|   no2_mo_avg_impute|    no_mo_avg_impute|pm25_mo_avg_impute|    assembler_output|       scaler_output| array_scaler_output|so

In [ ]:
# Run this cell
from pyspark.ml.evaluation import RegressionEvaluator

evaluator = RegressionEvaluator(labelCol='o3_mo_avg_impute', predictionCol='prediction', metricName='rmse')
rmse = evaluator.evaluate(df_prediction)

print('Root Mean Squared Error (RMSE): ', rmse)
print('Coefficients: ', model.coefficients)
print('Intercept: ', model.intercept)

Root Mean Squared Error (RMSE):  0.0044699054420452924
Coefficients:  [-0.0009113019162045537,-0.006104331381123103,0.004131654946987257]
Intercept:  0.0323286074969465


# *Explanation on the Required Challenge:*

I used the similar code prompt from this notebook and replaced the required values such as the `o3_mo_avg_impute` and `pm10_mo_avg_scale` in the `featuresCol=features, labelCol=o3_mo_avg_impute` using the linear regression model, respectively.

Furthermore, the `RegressionEvaluator` to calculate the Root Mean Squared Error of the model, which measures the average deviation of predictions from the model and actual values of `o3_mo_avg_impute`. The lowest RMSE, the better a fit to data. I also printed the model's coefficients and intercepts, enabling me to see how much each feature contributed to a prediction.

Hence, it displayed the following values for the challenge:

* Coefficients: `[-0.000911, -0.006104, 0.004132]` Coefficients are the effect of each feature, `co_mo_avg_scale`, `nox_mo_avg_scale`, and `pm10_mo_avg_scale` to the predictive `o3_mo_avg_impute`. A positive coefficient means that if the feature increases, the target value increases; otherwise, a negative coefficient means that if the feature increases, the target value decreases.

* Intercept: `0.0323` – This is the predicted value of `o3_mo_avg_impute` when all feature values are zero.